python/raydp/spark/dataset.py

```diff
@@ -449,3 +455,59 @@ def create_ml_dataset_from_spark(df: sql.DataFrame,
                                  node_hints: List[str] = None) -> MLDataset:
     return RayMLDataset.from_spark(
         df, num_shards, shuffle, shuffle_seed, fs_directory, compression, node_hints)
+
+def spark_dataframe_to_ray_dataset(df: sql.DataFrame):
+    blocks, _ = _save_spark_df_to_object_store(df)
+    return from_arrow(blocks)
+
+def _convert_blocks_to_dataframe(blocks):
+    # connect to ray
+    if not ray.is_initialized():
+        # raise log level to avoid filling log with this
+        ray.init(address="auto", logging_level=logging.WARN)
+    for block in blocks:
+        dfs = []
+        for b in block["ref"]:
+            ref = ray.cloudpickle.loads(b)
+            data = ray.get(ref)
+            dfs.append(data.to_pandas())
+        yield pd.concat(dfs)
+
+def _convert_by_udf(spark: sql.SparkSession,
+                    blocks: List[ObjectRef],
+                    schema: StructType) -> DataFrame:
+    s = StructType([StructField("ref", BinaryType(), False)])
+    ref_list = [(ray.cloudpickle.dumps(block),) for block in blocks]
+    blocks_df = spark.createDataFrame(ref_list, schema)
+    return blocks_df.mapInPandas(_convert_blocks_to_dataframe, schema)
+
+def _convert_by_rdd(spark: sql.SparkSession, blocks: Dataset, schema: StructType) -> DataFrame:
+    object_ids = []
+    locations = []
+    for block in blocks:
+        # address is needed for locality
+        locations.append(ray.worker.global_worker.core_worker.get_owner_address(block))
```

Comment:
Will `get_owner_address` return multiple locations if the block is stored in more than one node?

Reply:
This isn't a public API of Ray. If you really need it, you should call it from ray/data code and pass in the addresses here as an optional hint.

That way raydp isn't depending on private Ray APIs.

Reply:
I have added a developer api to ray. For now addresses are needed to register ownership in java.